encoding/template/run_length_encode.go

```diff
@@ -0,0 +1,80 @@
+package template
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/mauricelam/genny/generic"
+	"github.com/xichen2020/eventdb/x/bytes"
+)
+
+// GenericValue is a generic type.
+type GenericValue generic.Type
+
+// ValueMarshalFn reads a GenericValue from an `io.Reader`.
+type ValueMarshalFn func(writer io.Writer, value GenericValue) error
+
+// ForwardValueIterator allows iterating over a stream of GenericValue.
+type ForwardValueIterator interface {
+	generic.Type
+
+	io.Closer
+	Next() bool
+	Err() error
+	Current() GenericValue
+}
+
+// runLengthEncodeValue run length encodes a stream of GenericValue.
+func runLengthEncodeValue(
+	writer io.Writer,
+	extBuf *[]byte, // extBuf is an external byte buffer for memory re-use.
+	marshalFn ValueMarshalFn,
+	valuesIt ForwardValueIterator,
+) error {
+	// Ensure that our buffer size is large enough to handle varint ops.
+	*extBuf = bytes.EnsureBufferSize(*extBuf, binary.MaxVarintLen64, bytes.DontCopyData)
+
+	var (
+		last        *GenericValue
+		repetitions int
+	)
+	for valuesIt.Next() {
+		curr := valuesIt.Current()
+		// Set on the first value.
+		if last == nil {
+			last = &curr
+			continue
+		}
+
+		// Incrememnt repetitions and continue if we find a repetition.
+		if *last == curr {
+			repetitions++
+			continue
+		}
+
+		// last and curr don't match, write out the run length encoded repetitions
+		// and perform housekeeping.
+		n := binary.PutVarint(*extBuf, int64(repetitions))
```

Comment:
So if a value only appears once, the `repetitions` is 0 instead of 1? Conventionally the `repetitions` should be the same as the number of times value appears so I'd think `repetitions` should rather be 1 here?

Reply:
I was thinking that it should be 0, as it's the # of times a value is repeated. And if it shows up only once, that means there are 0 repetitions.

I can change it to 1 here though.